src/Microsoft.ML.OnnxTransform/OnnxUtils.cs

```diff
@@ -98,239 +102,151 @@ public static OnnxModel CreateFromBytes(byte[] modelBytes)
             // or keep the dir/file and write proper cleanup when application closes
         }
 
-        public List<Tensor> Run(List<Tensor> inputTensors)
+        /// <summary>
+        /// Uses an open session to score a list of NamedOnnxValues.
+        /// </summary>
+        /// <param name="inputNamedOnnxValues">The NamedOnnxValues to score</param>
+        /// <returns>Resulting output NamedOnnxValues list</returns>
+        public List<NamedOnnxValue> Run(List<NamedOnnxValue> inputNamedOnnxValues)
         {
-            var outputTensors = _modelManager.RunModel(
-                _modelName, _ignoredVersion, InputNames, inputTensors, OutputNames);
-
-            return outputTensors;
+            var outputNameOnnxValues = _session.Run(inputNamedOnnxValues);
+            var results = new List<NamedOnnxValue>();
+            foreach (var value in outputNameOnnxValues)
+            {
+                results.Add(value);
+            }
+            return results;
         }
 
+        /// <summary>
+        /// Convert the model to a byte array.
+        /// </summary>
+        /// <returns>byte[]</returns>
         public byte[] ToByteArray()
         {
             return File.ReadAllBytes(_modelFile);
         }
 
-        private OnnxNodeInfo[] GetInputsInfo()
-        {
-            return DictToNodesInfo(
-                    _modelManager.GetInputTypeDict(_modelName, _ignoredVersion),
-                    _modelManager.GetInputShapesDict(_modelName, _ignoredVersion));
-        }
-
-        private OnnxNodeInfo[] GetOutputsInfo()
+        /// <summary>
+        /// Returns input metadata of the ONNX model.
+        /// </summary>
+        /// <returns>OnnxNodeInfo[]</returns>
+        public OnnxNodeInfo[] GetInputsInfo()
         {
-            return DictToNodesInfo(
-                    _modelManager.GetOutputTypeDict(_modelName, _ignoredVersion),
-                    _modelManager.GetOutputShapesDict(_modelName, _ignoredVersion));
+            var nodeInfos = new List<OnnxNodeInfo>();
+            var inputMeta = _session.InputMetadata;
+            foreach (var kv in inputMeta)
+            {
+                nodeInfos.Add( new OnnxNodeInfo(kv.Key, kv.Value.Dimensions.ToList(), kv.Value.ElementType));
+            }
+            return nodeInfos.ToArray();
```

Comment:
Why call `ToArray()` here and copy the values out and create another object?

Instead, just return `IReadOnlyList<OnnxNodeInfo>`, and have the caller use that directly.

Reply:
reduced to  a 1-liner returning IEnumerable<OnnxNodeInfo>.